src/utils/type-safety.ts

```diff
@@ -0,0 +1,8 @@
+import { ZodSchema } from "zod";
+
+import { Exactly } from "./types";
+
+export const EnsureSchema: <T>() => <U>(
+  u: Exactly<ZodSchema<T>, ZodSchema<U>> extends true ? ZodSchema<U> : never
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-explicit-any
+) => ZodSchema<T> = () => (u) => u as any;
```

Comment:
Good job! It covers all the strictness checks that we were missing. I've improved the return type of the function to have a better match with the `ZodSchema` interface instead of `ZodType`. At the end the last one is just an alias of the first one, but it's the one that we use in the wallet. I've also removed the 2nd and 3d generic types as they are not needed.

```ts
export const EnsureSchema: <T>() => <U>(
  u: Exact<ZodSchema<T>, ZodSchema<U>> extends true
    ? Exact<Required<T>, Required<U>> extends true
      ? ZodSchema<U>
      : never
    : never
) => ZodSchema<T> =
  () =>
  <T>(u: unknown) =>
    u as ZodSchema<T>;

```

Reply:
We've also agreed in using the `Required` native helper instead of the `RemoveOptional` suggestion.